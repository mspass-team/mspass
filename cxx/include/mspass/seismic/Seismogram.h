#ifndef _SEISMOGRAM_H_
#define _SEISMOGRAM_H_
#include "mspass/seismic/CoreSeismogram.h"
#include "mspass/utility/ProcessingHistory.h"
#include "mspass/utility/ErrorLogger.h"

namespace mspass::seismic{
/*! \brief Implemntation of Seismogram for MsPASS.

This is the working version of a three-component seismogram object used
in the MsPASS framework.   It extends CoreSeismogram by adding
ProcessingHistory.   */
class Seismogram : public mspass::seismic::CoreSeismogram,
   public mspass::utility::ProcessingHistory
{
public:
  /*! Default constructor.   Only runs subclass default constructors. */
  Seismogram() : mspass::seismic::CoreSeismogram(),mspass::utility::ProcessingHistory(){};
  /*! Bare bones constructor allocates space and little else.

  Sometimes it is helpful to construct a skeleton that can be fleshed out
  manually.   This constructor allocates a 3xnsamples array and sets the
  matrix to all zeros.  It also sets the orientation information to cardinal
  and defines the data in relative time units.  The data are marked dead
  because the assumption is the caller will fill out commonly needed basic
  Metadata, load some kind of sample data into the data matrix, and then
  call the set_live method when that process is completed.   That kind of
  manipulation is most common in preparing simulation or test data where
  the common tags on real data do not exist and need to be defined manually
  for the simulatioon or test.   The history section is initialized with
  the default constructor, which currently means it is empty.   If a simulation
  or test requires a history origin the user must load it manaually.

  \param nsamples is the number of 3c vector samples to iniitalize the
    object with (creates a 3xnsamples matrix).
    */
  Seismogram(const size_t nsamples);
  /*! \brief Construct from lower level CoreSeismogram.

  In MsPASS CoreSeismogram has the primary functions that define the
  concept of a three-component seismogram.   Seismogram implements
  mspass specific features needed to mesh with the mspass processing system.
  This constructor clones onlyh the CoreSeismogram components and initializes
  the ProcessingHistory with the default constructor leaves the history
  in an empty state.   Users should call methods in ProcessingHistory to
  initiate a valid history chain.   Processing can continue if left in
  that state, but the history chain will have an undefined origin.  Job
  information will also be lost if not initialized (see BasicProcessingHistory)

  \param d is the data to be copied to create the new Seismogram object
   */
  Seismogram(const mspass::seismic::CoreSeismogram& d);
  /*! Contruct from a core seismogram and initialize history as origin.

  This constructor is a variant of a similar one built only from a
  CoreSeismogram.  This constuctor is intended to be used mainly on
  simulation data created by some mechanism (e.g. a python procedure).
  It is more rigid that the simple one arg constructor as it will
  create a top level history record.   The record will mark the result
  as an origin with an id set as a uuid created by a random number
  generator (boost implementation).  The jobname and jobid are
  frozen as "test". Use a different constructor and/or reset job info
  if more flexibility is desired.  Use of this constructor
  is recommended only for test python programs that do not need to
  interact with MongoDB.

  \param is core data to be cloned
  \param alg is the algorithm name to set for the origin history record.
  */
  Seismogram(const mspass::seismic::CoreSeismogram& d, const std::string alg);
  /*! \brief Partial constructor from base classes.

  This is a partial constructor useful sometimes for building a Seismogram
  object from scratch such as in a simultion.  It clones the Metadata and
  uses attributes in BasicTimeSeries to build a framework that data can
  be added into.  It initialzies the data array to npts stored in the
  BasicTimeSeries passed to the constructor.

  \param bts is the BasicTimeSeries that overrides any md data.
  \param md is the Metadata cloned to make the partially constructed object.
  */
  Seismogram(const mspass::seismic::BasicTimeSeries& bts,
      const mspass::utility::Metadata& md);


  /*! \brief Construct from all pieces.

This constructor build a Seismogram object from all the pieces that define
this highest level object in mspass.   This constructor is planned to
be hidden from python programmers and not exposed with pybind11 wrappers
because is has some potentially undesirable side effects if not used
carefully.  The primary purpose of this constructor is for serialization
and deserializaton in spark with pickle.  The pickle interface is
purely in C for this function so again python programmers don't need
to see this constructor.  The parameter names are obvious because
they are associated one for one with the objects they are used to
construct.  The only exceptions are card and ortho which are booleans
used to set the internal booleans components_are_cardinal and
components_are_orthogonal respectively (two internals users should not
mess with).   tm is also a dmatrix representation the tmatrix stored
internally as a 2d C array, but we use the dmatrix to mesh
with serialization.
*/
  Seismogram(const mspass::seismic::BasicTimeSeries& b, const mspass::utility::Metadata& m,
    const mspass::utility::ProcessingHistory& his,
      const bool card, const bool ortho,
        const mspass::utility::dmatrix& tm, const mspass::utility::dmatrix& uin);
  /*! Constructor driven by a Metadata object.

  The flexibilityof Metadata makes it helpful at times to build a Seismogram
  object driven by Metadata key:value pairs.   This implementation always uses
  file io to read the actual data and clone the Metadata.   ProcessingHistory
  is initialized to define the result as raw data.  The id field is set from
  Metadata using the special (frozen) key "wf_id" that is assumed to hold the
  string representation of an ObjectID of the waveform read.

  Note all but the Metadata argument are defaulted, but setting the jobname
  and jobid is strongly recommended to avoid ambiguity - these define a unique
  run of a particular workflow.


  \param md is the Metadata that is used to drive the constructor.
  \param jobname is the jobname (defaults to test, but should be changed)
  \param readername is the algorithm name assigned to the top level history
     record.   Defaults to "load3C"
  */
  Seismogram(const Metadata& md,const std::string jobname=std::string("test"),
    const std::string jobid=std::string("UNDEFINED"),
      const std::string readername=std::string("load3C"),
        const std::string algid=std::string("0"));

/*! \brief Construct from Metadata definition that includes data path.
 *
 A Metadata object is sufficiently general that it can contain enough
 information to contruct an object from attributes contained in it.
 This constuctor uses that approach, with the actual loading of data
 being an option (on by default).   In mspass this is constructor is
 used to load data with Metadata constructed from MongoDB and then
 using the path created from two parameters (dir and dfile used as
 in css3.0 wfdisc) to read data.   The API is general but the
 implementation in mspass is very rigid.   It blindly assumes the
 data being read are binary doubles in the right byte order and
 ordered in the native order for dmatrix (Fortran order).  i.e.
 the constuctor does a raw fread of ns*3 doubles into the internal
 array used in the dmatrix implementation.

 A second element of the Metadata that is special for MsPASS is the
 handling of the transformation matrix by this constructor.   In MsPASS
 the transformation matrix is stored as a python object in MongoDB.
 This constructor aims to fetch that entity with the key 'tmatrix'.
 To be more robust and simpler to use with data not loaded from mongodb
 we default tmatrix to assume the data are in standard coordinates.  That is,
 if the key tmatrix is not defined in Metadata passed as arg0, the
 constructor assumes it should set the transformation matrix to an identity.
 Use set_transformation_matrix if that assumption is wrong for your data.

 \param md is the Metadata used for the construction.

 \param load_data if true (default) a file name is constructed from
 dir+"/"+dfile, the file is openned, fseek is called to foff,
 data are read with fread, and the file is closed.  If false a dmatrix
 for u is still created of size 3xns, but the matrix is only initialized
 to all zeros.

 \exception  Will throw a MsPASSError if required metadata are missing.
 */
   Seismogram(const Metadata& md, bool load_data)
     : mspass::seismic::CoreSeismogram(md,load_data), mspass::utility::ProcessingHistory()
   {};
  /*! Standard copy constructor. */
  Seismogram(const Seismogram& parent)
    : mspass::seismic::CoreSeismogram(parent), mspass::utility::ProcessingHistory(parent)
  {};
  virtual ~Seismogram(){};
  /*! Standard assignment operator. */
  Seismogram& operator=(const Seismogram& parent);
  /*! \brief Load just the ProcessingHistory data from another data source.

  Some algorithms don't handle processing history.   In those situations it
  can prove helpful to manage ProcessingHistory separately and load the
  history data through this method.

  \param h is the ProcessingHistory data to copy into this Seismogram.
  */
  void load_history(const mspass::utility::ProcessingHistory& h);
  /*! Return an estimate of the memmory use by the data in this object.

  Memory consumed by a Seismogram object is needed to implement the
  __sizeof__ method in python that dask/spark use to manage memory.  Without
  that feature we had memory fault issues.  Note the estimate this
  method returns should not be expected to be exact.  The MsPASS implementation
  or any alternative implementation avoids an exact calculation because it
  requries an (expensive) traversal of multiple map containers.
  */
  size_t memory_use() const;
};
}//END mspass::seismic namespace
#endif
