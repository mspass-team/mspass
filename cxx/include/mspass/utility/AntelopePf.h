#ifndef _ANTELOPE_PF_H_
#define _ANTELOPE_PF_H_

#include <list>
#include <string>
#include "mspass/utility/Metadata.h"
namespace mspass{
namespace utility{
/*! key for accessing Tbl and Arr entries.
 *
 These are of use only when this object is converted to a Metadata
 object using the ConvertToMetadata method.   This object holds
 a representation of Tbl and Arr sections of the input pffile as 
 in map containers that are converted by that method and stored 
 with the boost::any capability.   This allows them to be retrieved if
 desired with these keys and the correct type specification.   
 */
const std::string pftbl_key("AntelopePfTbl");
const std::string pfarr_key("AntelopePfArr");
/*! \brief C++ object version of a parameter file.

   This object encapsulates the Antelope concept of a parameter
file in a single wrapper.   The main constructor is actually
act much like the Antelope pfread procedure.
   Internally this object does not use an antelope Pf at all
directly, but it is a child of Metadata.  Simple attributes
(i.e. key-value pairs) are posted directly to the Metadata 
associative array (map) container.  Note that the
parser attempts to guess the type of each value given in the
obvious ways (periods imply real numbers, e or E imply real
numbers, etc.) but the algorithm used may not be foolproof. The get
methods are from the Metadata object.  Be warned like the Metadata 
object the type of an entry for a key can change and will be the last
one set.   
    An Antelope Tbl in a pf file is converted to an stl list
container of stl::string's that contain the input lines.  This is
a strong divergence from the tbl interface of Antelope, but one
I judged a reasonable modernization.   Similarly, Arr's are
converted to what I am here calling a "branch".   Branches
are map indexed containers with the key pointing to nested
versions of this sampe object type. This is in keeping with the way
Arr's are used in antelope, but with an object flavor instead of
the pointer style of pfget_arr.   Thus, get_branch returns a
AntelopePf object instead of a pointer that has to be memory
managed.
    A final note about this beast is that the entire thing was
created with a tacit assumption the object itself is not huge.
i.e. this implementation may not scale well if applied to very
large (millions) line pf files.  This is a job for something like 
MongoDB.   View this as a convenient format for building a Metadata
object.  Note, a code fragment to use this to create lower level
metadata would go like this:
  AntelopePf pfdata("example");   //parameter file constructor
  Metadata md(dynamic_cast<Metadata&>(pfdata);

\author Gary L. Pavlis, Indiana University

This version for MsPASS is derived from a similar thing called a 
PfStyleMetadata object in antelope contrib.   
*/
class AntelopePf : public Metadata
{
public:
    /*! Default constructor - does nothing.*/
    AntelopePf():Metadata(){};
    /*! \brief Construct from a base pf name.

    This constructor acts like the antelope pfread function for
    parameter files constructed in plain C.   That is, the argument is
    a base name sans the .pf extension.   Like antelope's pfread it
    will follow the chain of directories defined by PFPATH.   As with
    the antelope pfread procedure the last file read takes precendence.
    Note if PFPATH is not defined the path defaults to ".".
    Further if pfbase begins with a slash (i.e. "/") it is assumed to be
    the actual file name to read and the PFPATH feature is disabled. 

      \param pfbase is a the base pf name (always adds .pf to each name in PFPATH)

      \exception - throws a MsPASSError object with an informative message if
        constuctor fails for any reason.
        */
    AntelopePf(std::string pfbase);
    /*! \brief Construct from a set of text lines.

      This is a cruder constructor that could be used for small pf file.
      Read the contents of a pf file into a vector of lines with each 
      line being one line from the pf.  The constructor then sequentially parses
      this series of lines as it would if reading from a file.

      \param lines is the modified version of the text pf file.*/
    AntelopePf(std::list<std::string> lines);
    /*!  Standard copy constructor. */
    AntelopePf(const AntelopePf& parent);
    /*! \brief get a Tbl component by key.

      Antelope has the idea of a tbl, which is a list of
      lines that are parsed independently.  This is the
      get method to extract one of these by its key.

      \param key is the key for the Tbl desired.
      \exception AntelopePfError will be thrown if the key
         is not present. */
    std::list<std::string> get_tbl(const std::string key) const;
    /*! \brief used for subtrees (nested Arrs in antelope)

       This method is used for nested Arr constructs.
       This returns a copy of the branch defined by key
       attached to this object.   The original from the parent
       is retained.

       \param key is the key used to locate this branch.
       \returns a copy of the contents of the branch linked to key.

       \exception AntelopePfError will be thrown if the key is not
            found. */
    AntelopePf get_branch(const std::string key) const;
    /*! Return a list of keys for branches (Arrs) in the pf file. */
    std::list<std::string> arr_keys() const;
    /*! Return a list of keys for Tbls in the pf.*/
    std::list<std::string> tbl_keys() const;
    /*! \brief Return an object with only simple name:value pairs.
     *
     The Metadata parent of this object only handles name:value pairs.
     The values can, however, be any object boost::any can handle.  
     The documentation says that means it is copy constructable.  
     For now this method returns an object containin the boost::any 
     values.   Any Arr and Tbl entries are pushed directly to the 
     output Metadata using boost::any and the two keys defined as 
     const strings at the top of this file (pftbl_key and pfarr_key).   

     \return Metadata sans any Arr and Tbl entries.
     */
    Metadata ConvertToMetadata();
    /*! Standard assignment operator, */
    AntelopePf& operator=(const AntelopePf& parent);
    /*! \brief save result in a pf format.

       This is functionally equivalent to the Antelope pfwrite
       procedure, but is a member of this object.   A feature of
       the current implementation is that all simply type parameters
       will usually be listed twice in the output file.   The reason
       is that the constructor attempts to guess type, but to allow
       for mistakes all simple parameters are also treated as string
       variables so get methods are more robust.

       \param ofs is a std::ostream (e.g. cout) where the result
          will be written in pf style.   Usually should end in ".pf".
          */
    void pfwrite(std::ostream& ofs);
private:
    std::map<std::string,std::list<std::string> > pftbls;
    /* This is used for nested Arrs */
    std::map<std::string, AntelopePf> pfbranches;
    /* This method is used to implemetn PFPATH - it is called for
    each secondary file in the PFPATH chain.   Returns the number of
    items changed */
    int merge_pfmf(AntelopePf& m);
};
/*! \brief Error class for AntelopePf object.

  This error object is similar to that for Metadata but tags
  all errors cleanly as originating from this child of Metadata.
  Note MsPASSError is a child of std::exception, so catch that
  to most easily fetch messages coming from this beast.
  */
class AntelopePfError : public MsPASSError
{
    public:
        AntelopePfError(){
            message=std::string("AntelopePfError->undefined error");
        };
        AntelopePfError(std::string mess)
        {
            message="AntelopePfError object message="+mess;
        };
        AntelopePfError(const char *mess)
        {
            message=std::string("AntelopePfError object message=")+mess;
        };
        void log_error(){
            std::cerr << message<<std::endl;
        };
};
/* Procedural functions using AntelopePf object */

/*! \brief Build a MetadataList using AntelopePf object.

A convenient format to define a MetadataList is an Antelope Pf file.  
This procedure creates a MetadataList from an AntelopePf object 
(generated from a pf file) keying on tag that defines a pf Tbl& 
section. 

\param m is the PfStyleMetadata object where you expect to find the list.
\param tag is the unique tag on the Tbl in the original Pf containing the
  data defining the MetadataList.

  */
MetadataList get_mdlist(const mspass::utility::AntelopePf& m, const std::string tag);
/*! Read a single pf file and return as an AntelopePf object. */
AntelopePf pfread(const std::string fname);

} // End utility namespace
} // End mspass namespace declaration
#endif
