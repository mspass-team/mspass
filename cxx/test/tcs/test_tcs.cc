#include <fstream>
//#include <boost/config.hpp>
//#include <boost/archive/text_oarchive.hpp>
//#include <boost/archive/text_iarchive.hpp>
#include "mspass/utility/MsPASSError.h"
#include "mspass/seismic/Seismogram.h"
#include "mspass/utility/AntelopePf.h"
using namespace std;
using namespace mspass::utility;
using namespace mspass::seismic;
int main(int argc, char **argv)
{
    string pfname("test_tcs");
    string amname="css3.0";
    //AntelopePf pf(pfname);
    ios::sync_with_stdio();
   try
   {
	int i,j;
	cout << "Testing simple constructors" << endl;
	CoreSeismogram s1;
        cout << "Success"<<endl;
        cout << "Testing space allocating constructor"<<endl;
	CoreSeismogram s2(100);
        cout << "Success"<<endl;
	// Initialize the contents of s2 and apply a rotation
	// matrix
	cout << "trying rotation" << endl;
	for(i=0;i<3;++i)
		for(j=0;j<100;++j)
		{
			if(i==2) s2.u(i,j)=1.0;
			else
				s2.u(i,j)=0.0;
		}
  s2.set_live();
  s2.set_npts(100);
  s2.set_t0(0.0);
	SphericalCoordinate sc;
	sc.phi=0.0;
	sc.theta=M_PI_4;
	s2.rotate(sc);
	cout << "one sample of (0,0,1) rotated by phi=0, theta=45 deg:"
		<< s2.u(0,0) << ", "
		<< s2.u(1,0) << ", "
		<< s2.u(2,0) << endl;
        cout << "Testing operator[] for sample 0"<<endl;
        vector<double> tvec=s2[0];
        cout <<"Success - here is the result (should be identical to above) which was extracted raw"<<endl;
        cout << tvec[0]<<", "<<tvec[1]<<", "<<tvec[2]<<endl;
        cout << "Testing with the overloaded method for operator[double]"
            <<endl;
        double testtime(0.0);
        tvec=s2[testtime];
        cout << "Result - should be identical to previous"<<endl;
        cout << tvec[0]<<", "<<tvec[1]<<", "<<tvec[2]<<endl;
	cout << "Restoring (tests rotate_to_standard method)" << endl;
	s2.rotate_to_standard();
	cout << "One sample of restored (0,0,1): "
		<< s2.u(0,0) << ", "
		<< s2.u(1,0) << ", "
		<< s2.u(2,0) << endl;
        cout << "Trying theta=45 deg and phi=-45 deg "<<endl;
        sc.phi=-M_PI_4;
        double *nu1=SphericalToUnitVector(sc);
        s2.u(0,0)=nu1[0];  s2.u(1,0)=nu1[1];  s2.u(2,0)=nu1[2];
        delete [] nu1;
        s2.rotate(sc);
	cout << "one sample of unit vector pointing  phi=-45, theta=45 deg after rotation:  "
		<< s2.u(0,0) << ", "
		<< s2.u(1,0) << ", "
		<< s2.u(2,0) << endl;
        cout << "Should be approximately (0,0,1)"<<endl;
	cout << "Restoring to standard" << endl;
	s2.rotate_to_standard();
	cout << "One sample of restored phi-45 theta=45 vector:  "
		<< s2.u(0,0) << ", "
		<< s2.u(1,0) << ", "
		<< s2.u(2,0) << endl;

	CoreSeismogram s3(100);
	for(i=0;i<3;++i)
		for(j=0;j<100;++j)
		{
			if(i==0) s3.u(i,j)=1.0;
			else
				s3.u(i,j)=0.0;
		}
  s3.set_live();
  s3.set_npts(100);
  s3.set_t0(0.0);
	// set some other vectors used in test below
	s3.u(0,1)=1.0;
	s3.u(1,1)=1.0;
	s3.u(2,1)=1.0;

	s3.u(0,2)=1.0;
	s3.u(1,2)=1.0;
	s3.u(2,2)=0.0;

	s3.u(0,3)=0.0;
	s3.u(1,3)=0.0;
	s3.u(2,3)=1.0;

	cout << "Trying unit vector rotation to (1,1,1) with (1,0,0) data"<<endl;
	double nu[3]={sqrt(3.0)/3.0,sqrt(3.0)/3.0,sqrt(3.0)/3.0};
	s3.rotate(nu);
        cout << "Transformation matrix created by this method:"<<endl;
        dmatrix tm=s3.get_transformation_matrix();
        cout << tm<<endl;
	cout << "one sample of (1,0,0) rotated to (1,1,1)"
		<< s3.u(0,0) << ", "
		<< s3.u(1,0) << ", "
		<< s3.u(2,0) << endl;
	cout << "one sample of (1,1,1) rotated to (1,1,1)"
		<< s3.u(0,1) << ", "
		<< s3.u(1,1) << ", "
		<< s3.u(2,1) << endl;
        cout << "Should be (0,0,sqrt(3))"<<endl;
	cout << "Restoring to standard" << endl;
	s3.rotate_to_standard();
	cout << "One sample of restored (1,0,0): "
		<< s3.u(0,0) << ", "
		<< s3.u(1,0) << ", "
		<< s3.u(2,0) << endl;
	cout << "One sample of restored (1,1,1): "
		<< s3.u(0,1) << ", "
		<< s3.u(1,1) << ", "
		<< s3.u(2,1) << endl;
	cout << "Trying spherical coordinate transformation to rotate phi=45 degrees,theta=0"
	<< endl;
	sc.phi=M_PI_4;
	sc.theta=0.0;
	s3.rotate(sc);
        cout << "Transformation matrix created by this method:"<<endl;
        tm=s3.get_transformation_matrix();
        cout << tm<<endl;
	cout << "one sample of transformed (1,0,0)"
		<< s3.u(0,0) << ", "
		<< s3.u(1,0) << ", "
		<< s3.u(2,0) << endl;
	cout << "One sample of transformed (1,1,1): "
		<< s3.u(0,1) << ", "
		<< s3.u(1,1) << ", "
		<< s3.u(2,1) << endl;
	cout << "one sample of transformed (1,1,0)"
		<< s3.u(0,2) << ", "
		<< s3.u(1,2) << ", "
		<< s3.u(2,2) << endl;
	cout << "one sample of transformed (0,0,1)"
		<< s3.u(0,3) << ", "
		<< s3.u(1,3) << ", "
		<< s3.u(2,3) << endl;
	s3.rotate_to_standard();
	cout << "Applying nonorthogonal transformation"<<endl;
	double a[3][3];
	a[0][0]=1.0;  a[0][1]=1.0;  a[0][2]=1.0;
	a[1][0]=-1.0;  a[1][1]=1.0;  a[1][2]=1.0;
	a[2][0]=0.0;  a[2][1]=-1.0;  a[2][2]=0.0;
	s3.transform(a);
	cout << "one sample of transformed (1,0,0):  "
		<< s3.u(0,0) << ", "
		<< s3.u(1,0) << ", "
		<< s3.u(2,0) << endl;
	cout << "One sample of transformed (1,1,1): "
		<< s3.u(0,1) << ", "
		<< s3.u(1,1) << ", "
		<< s3.u(2,1) << endl;
	cout << "one sample of transformed (1,1,0):  "
		<< s3.u(0,2) << ", "
		<< s3.u(1,2) << ", "
		<< s3.u(2,2) << endl;
	cout << "one sample of transformed (0,0,1):  "
		<< s3.u(0,3) << ", "
		<< s3.u(1,3) << ", "
		<< s3.u(2,3) << endl;
	cout << "Testing back conversion to standard coordinates"<<endl;
	s3.rotate_to_standard();
	cout << "One sample of restored (1,0,0): "
		<< s3.u(0,0) << ", "
		<< s3.u(1,0) << ", "
		<< s3.u(2,0) << endl;
	cout << "One sample of restored (1,1,1): "
		<< s3.u(0,1) << ", "
		<< s3.u(1,1) << ", "
		<< s3.u(2,1) << endl;
	cout << "One sample of restored (1,1,0): "
		<< s3.u(0,2) << ", "
		<< s3.u(1,2) << ", "
		<< s3.u(2,2) << endl;
	cout << "One sample of restored (0,0,1): "
		<< s3.u(0,3) << ", "
		<< s3.u(1,3) << ", "
		<< s3.u(2,3) << endl;

	cout << "Testing multiple, accumulated transformations" << endl;
	s3.rotate(M_PI_4);
	s3.transform(a);
        cout << "Accumulated transformation matrix from double transformation"
            <<endl;
        tm=s3.get_transformation_matrix();
        cout << tm;
	cout << "one sample of transformed (1,0,0):  "
		<< s3.u(0,0) << ", "
		<< s3.u(1,0) << ", "
		<< s3.u(2,0) << endl;
	cout << "One sample of transformed (1,1,1): "
		<< s3.u(0,1) << ", "
		<< s3.u(1,1) << ", "
		<< s3.u(2,1) << endl;
	cout << "one sample of transformed (1,1,0):  "
		<< s3.u(0,2) << ", "
		<< s3.u(1,2) << ", "
		<< s3.u(2,2) << endl;
	cout << "one sample of transformed (0,0,1):  "
		<< s3.u(0,3) << ", "
		<< s3.u(1,3) << ", "
		<< s3.u(2,3) << endl;
	s3.rotate_to_standard();
	cout << "One sample of restored (1,0,0): "
		<< s3.u(0,0) << ", "
		<< s3.u(1,0) << ", "
		<< s3.u(2,0) << endl;
	cout << "One sample of restored (1,1,1): "
		<< s3.u(0,1) << ", "
		<< s3.u(1,1) << ", "
		<< s3.u(2,1) << endl;
	cout << "One sample of restored (1,1,0): "
		<< s3.u(0,2) << ", "
		<< s3.u(1,2) << ", "
		<< s3.u(2,2) << endl;
	cout << "One sample of restored (0,0,1): "
		<< s3.u(0,3) << ", "
		<< s3.u(1,3) << ", "
		<< s3.u(2,3) << endl;

	/* We put this to metadata for verification below */
	s3.put_string("foo","bar");
        CoreSeismogram s4(s3);
        cout << "Testing free_surface_transformation method"<<endl;
	SlownessVector uvec;
	uvec.ux=0.17085;  // cos(-20deg)/5.5
	uvec.uy=-0.062185; // sin(-20deg)/5.5
	//s4.free_surface_transformation(uvec,8.0,3.5); // test exception
	s4.free_surface_transformation(uvec,5.0,3.5);
        cout << "Completed"<<endl;
        cout << "Computed transformation matrix:"<<endl;
        tm=s4.get_transformation_matrix();
        cout << tm<<endl;
	///
	cout << "Testing Seismogram constructor from CoreSeismogram"<<endl;
	Seismogram s5(s4);
	cout << "Transformation matrix - should be same as previous"<<endl;
	tm = s5.get_transformation_matrix();
	cout << tm<<endl;
	if(s5.live())
		cout << "Result is still marked live"<<endl;
	else
	{
		cout << "Result is marked dead - error"<<endl;
        	exit(-1);
	}
	cout << "Testing Seismogram copy constructor"<<endl;
	Seismogram s6(s5);
	if(s6.live())
		cout << "Result is still marked live"<<endl;
	else
	{
		cout << "Result is marked dead - error"<<endl;
        	exit(-1);
	}
	/* Test BasicTimeSeries was copied properly and verify 
	expected set metadata were copied */
	if( (s5.t0()!=s6.t0()) || (s5.npts()!=s6.npts()) || (s5.dt()!=s6.dt())
		|| (s5.timetype() != s6.timetype()) )
	{
		cout << "Error - copy constructor for Seismogram did not "
			<< "cleanly copy BasicTimeSeries"<<endl
			<< "Bug fix is required"<<endl;
		exit(-1);
	}
	try {
		string strtest;
		strtest=s5.get_string("foo");
		if(strtest!="bar")
		{
			cout << "Error - test metadata in Seismogram was garbled"<<endl;
			exit(-1);
		}
	}catch( MsPASSError& merr)
	{
		cout << merr.what()<<endl;
		cout << "Partial copy constructor for Seismogram from CoreSeismogram has a problem"<<endl;
		exit(-1);
	}
	try {
		string strtest;
		strtest=s6.get_string("foo");
		if(strtest!="bar")
		{
			cout << "Error - test metadata in Seismogram was garbled in Seismogram copy constructor"<<endl;
			exit(-1);
		}
	}catch( MsPASSError& merr)
	{
		cout << merr.what();
		cout << "Seismogram copy constructor has a problem"<<endl;
		exit(-1);
	}
	cout << "Copy constructor for Seismogram passed all tests"<<endl;
    }
    catch (MsPASSError&  serr)
    {
	serr.log_error();
        exit(-1);
    }
}
