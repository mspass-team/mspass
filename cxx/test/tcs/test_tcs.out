Testing simple constructors
Success
Testing space allocating constructor
Success
trying rotation
one sample of (0,0,1) rotated by phi=0, theta=45 deg:0, 0, 0
Testing operator[] for sample 0
Success - here is the result (should be identical to above) which was extracted raw
0, 0, 0
Testing with the overloaded method for operator[double]
Result - should be identical to previous
0, 0, 0
Restoring (tests rotate_to_standard method)
One sample of restored (0,0,1): 0, 0, 0
Trying theta=45 deg and phi=-45 deg 
one sample of unit vector pointing  phi=-45, theta=45 deg after rotation:  0, 0, 1
Should be approximately (0,0,1)
Restoring to standard
One sample of restored phi-45 theta=45 vector:  0.5, -0.5, 0.707107
Trying unit vector rotation to (1,1,1) with (1,0,0) data
Transformation matrix created by this method:
 0.707107 -0.707107 0
 0.408248 0.408248 -0.816497
 0.57735 0.57735 0.57735

one sample of (1,0,0) rotated to (1,1,1)0, 0, 0
one sample of (1,1,1) rotated to (1,1,1)1.11022e-16, 0, 1.73205
Should be (0,0,sqrt(3))
Restoring to standard
One sample of restored (1,0,0): 0, 0, 0
One sample of restored (1,1,1): 1, 1, 1
Trying spherical coordinate transformation to rotate phi=45 degrees,theta=0
Transformation matrix created by this method:
 0.707107 -0.707107 0
 0.707107 0.707107 -0
 0 0 1

one sample of transformed (1,0,0)0, 0, 0
One sample of transformed (1,1,1): 2.22045e-16, 1.41421, 1
one sample of transformed (1,1,0)1.11022e-16, 1.41421, 0
one sample of transformed (0,0,1)-3.92523e-17, 3.92523e-17, 1
Applying nonorthogonal transformation
one sample of transformed (1,0,0):  0, 0, 0
One sample of transformed (1,1,1): 3, 1, -1
one sample of transformed (1,1,0):  2, 0, -1
one sample of transformed (0,0,1):  1, 1, -5.55112e-17
Testing back conversion to standard coordinates
One sample of restored (1,0,0): 0, 0, 0
One sample of restored (1,1,1): 1, 1, 1
One sample of restored (1,1,0): 1, 1, 0
One sample of restored (0,0,1): 0, 5.55112e-17, 1
Testing multiple, accumulated transformations
Accumulated transformation matrix from double transformation
 1.41421 1.11022e-16 1
 -1.11022e-16 1.41421 1
 -0.707107 -0.707107 0
one sample of transformed (1,0,0):  0, 0, 0
One sample of transformed (1,1,1): 2.41421, 2.41421, -1.41421
one sample of transformed (1,1,0):  1.41421, 1.41421, -1.41421
one sample of transformed (0,0,1):  1, 1, -3.92523e-17
One sample of restored (1,0,0): 0, 0, 0
One sample of restored (1,1,1): 1, 1, 1
One sample of restored (1,1,0): 1, 1, 0
One sample of restored (0,0,1): 2.77556e-17, 8.32667e-17, 1
Testing free_surface_transformation method
Completed
Computed transformation matrix:
 -0.171012 -0.469846 0
 0.115793 -0.0421458 0.445447
 -0.597975 0.217647 0.228152

Testing Seismogram constructor from CoreSeismogram
Transformation matrix - should be same as previous
 -0.171012 -0.469846 0
 0.115793 -0.0421458 0.445447
 -0.597975 0.217647 0.228152

Result is still marked live
Testing Seismogram copy constructor
Result is still marked live
Copy constructor for Seismogram passed all tests
